### PR TITLE
chore(core)!: remove support for Node.js v16

### PR DESCRIPTION
**Pull request by legobeat**
Support for Node.js v16 is being removed for `lavamoat-core` and its dependencies:
- `lavamoat`-node
  - `lavamoat-browserify`
  - `lavamoat-perf`
  - `survey`
  - `@lavamoat/webpack`

This is a breaking change.

Node.js v16 support is still maintained for `@lavamoat/allow-scripts`.

#### Blocking
- #1192 
- #1165
- #1163
- #1218

### CODE REVIEW

File: .github/workflows/nodejs.yml [hunk 1]
```diff
@@ -17,15 +17,77 @@ jobs:
...
+              --workspace=packages/preinstall-always-fail
+              --workspace=packages/survey
+              --workspace=packages/tofu
+              --workspace=packages/webpack
+              --workspace=packages/yarn-plugin-allow-scripts
+
+          # TODO: Resolve https://github.com/LavaMoat/LavaMoat/issues/1211 and enable v20.x, v22.x
+          - node-version: 20.14.x
+            test-workspaces: >-
+              --workspace=packages/aa
```
Comment (legobeat) [inline .github/workflows/nodejs.yml:50]: The following workspaces are tested on latest 20.x/22.x as well as 20.14.x/22.1.x:

- `aa`
- `core`
- `tofu`

It seems redundant to double-test the others which support latest.

File: .github/workflows/nodejs.yml [hunk 2]
```diff
@@ -215,4 +218,7 @@ jobs:
           install-command: npm ci --foreground-scripts
       - name: Test
         run: |
+          npm run --if-present ${{ env.TEST_WORKSPACES }} build
+          npm run build:types
+          npm run --if-present ${{ env.TEST_WORKSPACES }} test:prep
```
Comment (legobeat) [inline .github/workflows/nodejs.yml:222]: Without this, we get:

```
> @lavamoat/lavapack@6.1.4 test
> ava


  ✔ sourcemap-regex › sourceMapDropperRegex detects embedded sourcemap
  ✔ sourcemap-regex › sourceMapDropperRegex detects embedded sourcemap with multiline comment
  ✔ sourcemap-regex › sourceMapDropperRegex detects embedded sourcemap with multiline comment and code after it
  ✔ sourcemap-regex › sourceMapDropperRegex detects embedded sourcemap in code with multiple multiline comments
  ✔ sourcemap-regex › sourceMapDropperRegex detects embedded sourcemap in code with malformed multiline comments
  ✔ sourcemap-regex › sourceMapDropperRegex does not detect embedded sourcemap in code without sourcemap
  - [skip] sourcemaps › sourcemaps - adjust maps for wrapper
  ✘ [fail]: index › dummy test Error thrown in test
  ✘ [fail]: index › detect invalid module Error thrown in test
  ✘ [fail]: index › sourcemap test Rejected promise returned by test
  ─

  index › dummy test

  test/index.spec.js:9

   8: test('dummy test', (t) => {
npm ERR! Lifecycle script `test` failed with error: 
   9:   pack({                   
npm ERR! Error: command failed 
   10:     includePrelude: false, 
npm ERR!   in workspace: @lavamoat/lavapack@6.1.4 

npm ERR!   at location: /Users/runner/work/LavaMoat/LavaMoat/packages/lavapack 
  Error thrown in test:

  Error {
    code: 'ENOENT',
    errno: -2,
    path: '/Users/runner/work/LavaMoat/LavaMoat/packages/lavapack/src/runtime.js',
    syscall: 'open',
    message: 'ENOENT: no such file or directory, open \'/Users/runner/work/LavaMoat/LavaMoat/packages/lavapack/src/runtime.js\'',
  }

  › createPacker (src/pack.js:48:16)
  › test/index.spec.js:9:3



  index › detect invalid module

  test/index.spec.js:96

   95:   t.plan(2)                
   96:   const packStream = pack({
   97:     raw: true,             

  Error thrown in test:

  Error {
    code: 'ENOENT',
    errno: -2,
    path: '/Users/runner/work/LavaMoat/LavaMoat/packages/lavapack/src/runtime.js',
    syscall: 'open',
    message: 'ENOENT: no such file or directory, open \'/Users/runner/work/LavaMoat/LavaMoat/packages/lavapack/src/runtime.js\'',
  }
```

https://github.com/LavaMoat/LavaMoat/actions/runs/10087632654/job/27892086277?pr=1230#step:6:316

File: .github/workflows/nodejs.yml [hunk 3]
```diff
@@ -17,15 +17,76 @@ jobs:
...
-          #- 20.x
-          #- 22.x
+        inlude:
+          - node-version: 16.x
+            test-workspaces: >-
+              --workspace=packages/aa
+              --workspace=packages/allow-scripts
+              --workspace=packages/lavapack
+              --workspace=packages/preinstall-always-fail
+              --workspace=packages/tofu
```
Comment (naugtur) [inline .github/workflows/nodejs.yml:29]: I'm not sure why lavapack and tofu are on this list. Understand the rest.

File: .github/workflows/nodejs.yml [hunk 4]
```diff
@@ -107,7 +172,6 @@ jobs:
         with:
           useRollingCache: true
           install-command: npm ci --foreground-scripts
-      # Limit testing to packages supported to run under windows
```
Comment (naugtur) [inline .github/workflows/nodejs.yml:110]: (nit) This comment looks useful, it's explaining the why. Or is it no longer true?

File: .github/workflows/nodejs.yml (hunk 3 above)
Comment (legobeat) [reply .github/workflows/nodejs.yml:29]: All packages that have support indicated for nodejs v16 in `engines.node` are also tested for it.

Removing support for these two packages could be considered, but could also be done in a follow-up?

Currently intended scope for this PR is for `core` and dependendents and these two are separate.

File: .github/workflows/nodejs.yml (hunk 4 above)
Comment (legobeat) [reply .github/workflows/nodejs.yml:110]: Previously the pattern was only used for Windows tests. Now it is used for all suites, so it doesn't seem that useful here anymore?

File: .github/workflows/nodejs.yml [hunk 5]
```diff
@@ -17,15 +17,76 @@ jobs:
...
+              --workspace=packages/aa
+              --workspace=packages/allow-scripts
+              --workspace=packages/core
+              --workspace=packages/lavapack
+              --workspace=packages/laverna
+              --workspace=packages/perf
+              --workspace=packages/preinstall-always-fail
+              --workspace=packages/survey
+              --workspace=packages/tofu
+              --workspace=packages/yarn-plugin-allow-scripts
```
Comment (naugtur) [inline .github/workflows/nodejs.yml:89]: (could) If there's a way to do it, having a variable called ALL_WORKSPACES here would be cool looking.

**Review by naugtur — APPROVED** [review]

File: .github/workflows/nodejs.yml (hunk 5 above)
Comment (legobeat) [reply .github/workflows/nodejs.yml:89]: Agreed but I came up short with a reasonable way at least this round.

GHA does not support yaml anchors, unfortunately, and the npm cli syntax is what it is.

File: .github/workflows/nodejs.yml (hunk 3 above)
Comment (legobeat) [reply .github/workflows/nodejs.yml:29]: - #1256 
- #1257

Comment (naugtur) [reply .github/workflows/nodejs.yml:29]: Sure, follow-up is fine - especially that it'd be considered a breaking change. I am not aware of them being dependencies of any packages where we intentionally keep node 16 support.

**Pull request merged**